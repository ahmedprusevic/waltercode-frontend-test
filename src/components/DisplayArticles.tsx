import React from "react";
import { makeStyles, Button } from "@material-ui/core";
import NewsCard from "./NewsCard";
import Loader from "./Loader";
import { useDispatch } from "react-redux";

export type DisplayArticlesProps = {
	articles: IArticle[];
	loading: boolean;
	increasePageNum: () => void;
	fetchedAll: boolean;
};

export default function DisplayArticles({
	articles,
	loading,
	increasePageNum,
	fetchedAll,
}: DisplayArticlesProps) {
	const classes = useStyles();
	const dispatch = useDispatch();

	return (
		<>
			<div className={classes.root}>
				{articles.map((article: IArticle, idx: number) => (
					<NewsCard
						article={article}
						key={`${article.publishedAt}${idx}`}
						index={idx}
					/>
				))}
			</div>
			{loading ? (
				<Loader />
			) : (
				<div className={classes.buttonDiv}>
					<Button
						variant="contained"
						color="primary"
						onClick={() => dispatch(increasePageNum())}
						disabled={fetchedAll}
					>
						Load More
					</Button>
				</div>
			)}
		</>
	);
}

const useStyles = makeStyles({
	root: {
		display: "flex",
		flexWrap: "wrap",
		justifyContent: "center",
	},
	buttonDiv: {
		margin: "1rem",
		display: "flex",
		justifyContent: "center",
	},
});
