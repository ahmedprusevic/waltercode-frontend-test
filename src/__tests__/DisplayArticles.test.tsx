import { configure, mount, shallow } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import DisplayArticles, {
	DisplayArticlesProps,
} from "../components/DisplayArticles";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import * as reactRedux from "react-redux";
import { Button } from "@material-ui/core";

configure({ adapter: new Adapter() });

afterEach(() => {
	jest.clearAllMocks();
});

describe("Snapshot testing and unit testing", () => {
	const initialState = { articles: [] };
	const mockStore = configureStore();
	const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");

	it("renders with basic props", () => {
		const displayArticlesProps: DisplayArticlesProps = {
			articles: [
				{
					title: "Mocked Title",
					source: {},
					description: "Mocked Description",
				},
			],
			loading: false,
			increasePageNum: jest.fn(),
			fetchedAll: false,
		};

		const store = mockStore(initialState);
		const component = shallow(
			<Provider store={store}>
				<DisplayArticles {...displayArticlesProps} />
			</Provider>
		);
		expect(component).toMatchSnapshot();
	});

	it("calls dispatch on button press", () => {
		const fakeDispatch = jest.fn();
		useDispatchMock.mockReturnValue(fakeDispatch);
		const displayArticlesProps: DisplayArticlesProps = {
			articles: [
				{
					title: "Mocked Title",
					source: {},
					description: "Mocked Description",
				},
			],
			loading: false,
			increasePageNum: jest.fn(),
			fetchedAll: false,
		};

		const store = mockStore(initialState);

		const component = shallow(<DisplayArticles {...displayArticlesProps} />);

		console.log(component.debug());
		component.find(Button).simulate("click");

		expect(fakeDispatch).toBeCalledTimes(1);
	});
});
