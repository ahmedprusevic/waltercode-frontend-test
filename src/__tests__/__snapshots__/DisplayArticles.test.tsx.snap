// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing and unit testing renders with basic props 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <DisplayArticles
    articles={
      Array [
        Object {
          "description": "Mocked Description",
          "source": Object {},
          "title": "Mocked Title",
        },
      ]
    }
    fetchedAll={false}
    increasePageNum={[MockFunction]}
    loading={false}
  />
</ContextProvider>
`;
